[simplefsdp auto-bucketing] ir node runtime estimation

ruisizhang123 opened the pull request:
Computation Estimation:

-  BaseSchedulerNode/FusedSchedulerNode: Get the generated triton code and use [do_bench](https://github.com/pytorch/pytorch/blob/85111cd165f108ffabb4a90083d59d7a867ebd9f/torch/_inductor/codegen/triton.py#L4234) mode to benchmark the runtime.
- ExternKernelSchedulerNode: Realize faketensor into realtensor, and log the cuda event time.

Communication Estimation:

- NCCL estimation: estimate communication using NCCL's [ncclGroupSimulateEnd API](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/group.html?highlight=ncclsiminfo_t) from this [PR](https://github.com/pytorch/pytorch/pull/149343).
- Profiling-based estimation: Realize faketensor into realtensor, and log the cuda event time for performing the communication.

Note: By default, I use Profiling-based estimation to estimate communication. The reason is it gives better estimation performance in multi-node settings.

At the end, it aggregates the estimated runtime from all ranks, and take the median runtime as the final estimated results.  

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160282
* #158609
* #158321
* #158098
* #158097
* __->__ #157572



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/simple_fsdp/estimator.py:
```diff
@@ -0,0 +1,257 @@
+    process_group = c10d.distributed_c10d._get_default_group()
+
+    tensor_input = create_real_tensor(inputs.data.get_size(), inputs.data.get_dtype(), inputs.data.get_device())
+    tensor_output = create_real_tensor(kernel.layout.size, kernel.layout.dtype, kernel.layout.device)
```
one downside to allocating real tensors at compile time to do the comms estimation is that if we are already close to peak memory at the time that we do the estimation, we risk compile causing an OOM. IIRC we've seen similar problems before with inductor, since it allocates tensors to do autotuning.

Mostly just calling this out as a potential downside. Although maybe this is still the best option, given that your work showed this is noticeably more accurate than other forms of comms estimation? Runtime PGO is another option but it is a much bigger lift to set up. Or just brainstorming: maybe we can also avoid these allocations being problematic by either offloading other tensors to CPU, or e.g. doing the runtime estimation with small tensor shapes, and having a way to accurately estimate the runtime for larger shapes.

ruisizhang123 replied:
These tensors are deleted in L126. Do you think there might be potential memory overhead even after they are deleted? 👀

wconstab commented on torch/_inductor/simple_fsdp/estimator.py:
```diff
@@ -0,0 +1,257 @@
+            end_evt.record()
+            end_evt.synchronize()
+
+            torch.cuda.synchronize()
```
nit: doing cuda sync / event sync should _not_ be necessary to get timings from the event.  I think i'd suggest just removing them.

ruisizhang123 replied:
We will need end_evt.synchronize() to tell the end_event has finished. Otherwise, it will throw error. I think `torch.cuda.synchronize()` is not necessary here tho.

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]

ruisizhang123 committed:
Update on "[simplefsdp auto-bucketing] ir node runtime estimation"

Computation Estimation:

-  BaseSchedulerNode/FusedSchedulerNode: Get the generated triton code and use [do_bench](https://github.com/pytorch/pytorch/blob/85111cd165f108ffabb4a90083d59d7a867ebd9f/torch/_inductor/codegen/triton.py#L4234) mode to benchmark the runtime.
- ExternKernelSchedulerNode: Realize faketensor into realtensor, and log the cuda event time.

Communication Estimation:

- NCCL estimation: estimate communication using NCCL's [ncclGroupSimulateEnd API](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/group.html?highlight=ncclsiminfo_t) from this [PR](https://github.com/pytorch/pytorch/pull/149343).
- Profiling-based estimation: Realize faketensor into realtensor, and log the cuda event time for performing the communication.

Note: By default, I use Profiling-based estimation to estimate communication. The reason is it gives better estimation performance in multi-node settings.

At the end, it aggregates the estimated runtime from all ranks, and take the median runtime as the final estimated results.  




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ruisizhang123 committed:
Update

[ghstack-poisoned]
